### ISSUE TITLE
Incorrect IP Address Reported in Logs with annotations / v0.0.5

### ISSUE THREAD

**Issue by reefland** [kube-vip/kube-vip-cloud-provider#59]
kube-vip-cloud-provider: `v0.0.5` on K3s `v1.26.4+k3s1`.

I noticed the message of Deprecate `service.spec.loadBalancerIP` and testing changing to annotation.

Before starting, confirmed I had a deprecated warning:
```
I0514 17:46:49.075861       1 loadBalancer.go:89] syncing service 'apt-cacher-ng' (07d8d6e5-a608-40e5-bfbd-5097c02e4d10)
W0514 17:46:49.075881       1 loadBalancer.go:94] service.Spec.LoadBalancerIP is defined but annotations 'kube-vip.io/loadbalancerIPs' is not, assume it's a legacy service, updates its annotations
I0514 17:46:49.075971       1 event.go:294] "Event occurred" object="apt-cacher-ng/apt-cacher-ng" fieldPath="" kind="Service" apiVersion="v1" type="Normal" reason="EnsuringLoadBalancer" message="Ensuring load balancer"
I0514 17:46:50.267391       1 loadBalancer.go:89] syncing service 'apt-cacher-ng' (07d8d6e5-a608-40e5-bfbd-5097c02e4d10)
```
 
Within the Helm Chart for this application I changed to annotation:
```
            annotations:
              kube-vip.io/loadbalancerIPs: 192.168.10.243
            type: LoadBalancer
            # loadBalancerIP: 192.168.10.243
```            
Watched the Kube-VIP CloudProvider logs:
```
I0514 17:50:51.463391       1 loadBalancer.go:89] syncing service 'apt-cacher-ng' (07d8d6e5-a608-40e5-bfbd-5097c02e4d10)
I0514 17:50:51.463437       1 event.go:294] "Event occurred" object="apt-cacher-ng/apt-cacher-ng" fieldPath="" kind="Service" apiVersion="v1" type="Normal" reason="LoadbalancerIP" message="192.168.10.243 -> "
I0514 17:50:51.463487       1 event.go:294] "Event occurred" object="apt-cacher-ng/apt-cacher-ng" fieldPath="" kind="Service" apiVersion="v1" type="Normal" reason="EnsuringLoadBalancer" message="Ensuring load balancer"
I0514 17:50:51.479789       1 loadBalancer.go:208] no cidr config for namespace [apt-cacher-ng] exists in key [cidr-apt-cacher-ng] configmap [kubevip]
I0514 17:50:51.479802       1 loadBalancer.go:213] Taking address from [cidr-global] pool
I0514 17:50:51.491008       1 loadBalancer.go:170] Updating service [apt-cacher-ng], with load balancer IPAM address [192.168.10.241]
I0514 17:50:51.512616       1 event.go:294] "Event occurred" object="apt-cacher-ng/apt-cacher-ng" fieldPath="" kind="Service" apiVersion="v1" type="Normal" reason="LoadbalancerIP" message=" -> 192.168.10.241"
I0514 17:50:51.515449       1 loadBalancer.go:89] syncing service 'apt-cacher-ng' (07d8d6e5-a608-40e5-bfbd-5097c02e4d10)
```

Within the logs above I see two references to an incorrect IP address:  `192.168.10.241`, this belongs to a different LoadBalancer service in the CIDR pool:

```
$ kubectl get svc mosquitto-mqtt -n mosquitto

NAME             TYPE           CLUSTER-IP     EXTERNAL-IP      PORT(S)          AGE
mosquitto-mqtt   LoadBalancer   10.43.55.109   192.168.10.241   1883:30214/TCP   276d
```
It's respective logs:
```
I0514 17:53:36.964549       1 loadBalancer.go:213] Taking address from [cidr-global] pool
I0514 17:53:36.984608       1 loadBalancer.go:170] Updating service [mosquitto-mqtt], with load balancer IPAM address [192.168.10.241]
I0514 17:53:37.002935       1 event.go:294] "Event occurred" object="mosquitto/mosquitto-mqtt" fieldPath="" kind="Service" apiVersion="v1" type="Normal" reason="LoadbalancerIP" message=" -> 192.168.10.241"
I0514 17:53:37.003038       1 loadBalancer.go:89] syncing service 'mosquitto-mqtt' (cab480f2-8721-46c4-866a-99764b17a3cc)
```

In the end the service was assigned the correct IP address.  Not sure if this is just a logging issue?
```
$ kubectl get svc -n apt-cacher-ng

NAME            TYPE           CLUSTER-IP      EXTERNAL-IP      PORT(S)          AGE
apt-cacher-ng   LoadBalancer   10.43.171.146   192.168.10.243   3142:32669/TCP   150d
```

---
I observed this happening more than once, for example:
```
I0514 18:51:31.749840       1 loadBalancer.go:208] no cidr config for namespace [syncthing] exists in key [cidr-syncthing] configmap [kubevip]
I0514 18:51:31.749862       1 loadBalancer.go:213] Taking address from [cidr-global] pool
I0514 18:51:31.777835       1 loadBalancer.go:170] Updating service [syncthing-listen], with load balancer IPAM address [192.168.10.242]
I0514 18:51:31.796257       1 event.go:294] "Event occurred" object="syncthing/syncthing-listen" fieldPath="" kind="Service" apiVersion="v1" type="Normal" reason="LoadbalancerIP" message=" -> 192.168.10.242"
I0514 18:51:31.796674       1 event.go:294] "Event occurred" object="syncthing/syncthing-listen" fieldPath="" kind="Service" apiVersion="v1" type="Normal" reason="EnsuredLoadBalancer" message="Ensured load balancer"
I0514 18:51:31.796683       1 loadBalancer.go:89] syncing service 'syncthing-listen' (d756dca6-d0cd-4145-a569-4715c6b9fbae)
I0514 18:51:31.796730       1 event.go:294] "Event occurred" object="syncthing/syncthing-listen" fieldPath="" kind="Service" apiVersion="v1" type="Normal" reason="EnsuringLoadBalancer" message="Ensuring load balancer"
I0514 18:51:31.796747       1 event.go:294] "Event occurred" object="syncthing/syncthing-listen" fieldPath="" kind="Service" apiVersion="v1" type="Normal" reason="EnsuredLoadBalancer" message="Ensured load balancer"
I0514 18:51:36.331556       1 loadBalancer.go:89] syncing service 'syncthing-listen' (d756dca6-d0cd-4145-a569-4715c6b9fbae)
I0514 18:51:36.331691       1 event.go:294] "Event occurred" object="syncthing/syncthing-listen" fieldPath="" kind="Service" apiVersion="v1" type="Normal" reason="EnsuringLoadBalancer" message="Ensuring load balancer"
I0514 18:51:36.331754       1 event.go:294] "Event occurred" object="syncthing/syncthing-listen" fieldPath="" kind="Service" apiVersion="v1" type="Normal" reason="EnsuredLoadBalancer" message="Ensured load balancer"
```

This service was defined to use `192.168.10.244` but the logs above show `192.168.10.242`:
```
            annotations:
              kube-vip.io/loadbalancerIPs: 192.168.10.244
...
            type: LoadBalancer
            # loadBalancerIP: 192.168.10.244
```

The service is updated correctly:
```
metadata:
  annotations:
    kube-vip.io/loadbalancerIPs: 192.168.10.244
```

```
$ kubectl get svc syncthing-listen -n syncthing

NAME               TYPE           CLUSTER-IP      EXTERNAL-IP      PORT(S)                           AGE
syncthing-listen   LoadBalancer   10.43.224.251   192.168.10.244   21027:32291/UDP,22000:30639/TCP   140d
```

Seems to just be a logging issue as far as I can tell.

**Comment by lubronzhan** [comment]
Hi, I will take a look next week. I'm OOO this week

**Comment by lubronzhan** [comment]
Ok I'm back from the vacation. I think it's because I tries to set both kube-vip annotation `kube-vip.io/loadbalancerIPs` and spec.loadbalancerIP, and I only checked for the spec at this moment, not the annotation. It should also check for annotation as well.
Let me create a quick fix

**Comment by lubronzhan** [comment]
Hi @reefland would you mind review the PR as well? https://github.com/kube-vip/kube-vip-cloud-provider/pull/62

**Comment by reefland** [comment]
Tested 0.0.7, seemed to work, this was logged:
```
I0523 17:00:59.014424       1 loadBalancer.go:91] syncing service 'apt-cacher-ng' (07d8d6e5-a608-40e5-bfbd-5097c02e4d10)

I0523 17:00:59.014440       1 event.go:294] "Event occurred" object="apt-cacher-ng/apt-cacher-ng" fieldPath="" kind="Service" apiVersion="v1" type="Normal" reason="LoadbalancerIP" message="192.168.10.243 -> "

W0523 17:00:59.014458       1 loadBalancer.go:121] service 'apt-cacher-ng/apt-cacher-ng' annotations 'kube-vip.io/loadbalancerIPs' is defined but service.Spec.LoadBalancerIP is not. Assume it's not legacy service

I0523 17:00:59.014469       1 event.go:294] "Event occurred" object="apt-cacher-ng/apt-cacher-ng" fieldPath="" kind="Service" apiVersion="v1" type="Normal" reason="EnsuringLoadBalancer" message="Ensuring load balancer"
```

Not clear what this line is trying to tell me:
* `reason="LoadbalancerIP" message="192.168.10.243 -> "`

Am I correct in assuming we are supposed to remove the `loadBalancerIP: 192.168.10.243` from the service, and only use the annotation?  If so then why is this a warning? 
```
W0523 17:00:59.014458       1 loadBalancer.go:121] service 'apt-cacher-ng/apt-cacher-ng' annotations 'kube-vip.io/loadbalancerIPs' is defined but service.Spec.LoadBalancerIP is not. Assume it's not legacy service
```
I don't see it logging anything about the IP address specified in the annotations.

**Comment by lubronzhan** [comment]
It's just to notify it's new type of service.
You are supposed to only set the annotation.
Will remove that line in the future.

> Not clear what this line is trying to tell me:
> reason="LoadbalancerIP" message="192.168.10.243 -> "

this looks weird. Let me take a look

**Comment by lubronzhan** [comment]
Did you remove the spec.loadblancerIP from the service
https://github.com/kubernetes/cloud-provider/blob/master/controllers/service/controller.go#L559-L563
Cloud provider just log the this has been changed and might need to update the service.
But it should be ok, since if you only use annotation

**Comment by lubronzhan** [comment]
> I don't see it logging anything about the IP address specified in the annotations.

If you set static ip in the annotation by yourself, then kube-vip-cloud-provider won't do anything just add a label of implemenation. Kube-vip will advertise the ip

Unless you let kube-vip-cloud-provider to allocate a ip from the pool, it won't log anything.

Did you see any logging if you create a service with predefined `spec.loadblancerIP`

**Comment by lubronzhan** [comment]
Made some improvement here https://github.com/kube-vip/kube-vip-cloud-provider/pull/65

**Comment by reefland** [comment]
> Did you remove the spec.loadblancerIP from the service

Yes I did.  

```
$ k get svc apt-cacher-ng -n apt-cacher-ng -o yaml
apiVersion: v1
kind: Service
metadata:
  annotations:
    kube-vip.io/loadbalancerIPs: 192.168.10.243
    kube-vip.io/vipHost: k3s03
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"v1","kind":"Service","metadata":{"annotations":{"kube-vip.io/loadbalancerIPs":"192.168.10.243"},"labels":{"app.kubernetes.io/instance":"apt-cacher-ng","app.kubernetes.io/managed-by":"Helm","app.kubernetes.io/name":"apt-cacher-ng","app.kubernetes.io/version":"3.7.4-20220421","argocd.argoproj.io/instance":"apt-cacher-ng","helm.sh/chart":"apt-cacher-ng-0.0.5"},"name":"apt-cacher-ng","namespace":"apt-cacher-ng"},"spec":{"ports":[{"name":"http","port":3142,"protocol":"TCP","targetPort":3142}],"selector":{"app.kubernetes.io/instance":"apt-cacher-ng","app.kubernetes.io/name":"apt-cacher-ng"},"type":"LoadBalancer"}}
  creationTimestamp: "2022-12-14T21:27:23Z"
  finalizers:
  - service.kubernetes.io/load-balancer-cleanup
  labels:
    app.kubernetes.io/instance: apt-cacher-ng
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/name: apt-cacher-ng
    app.kubernetes.io/version: 3.7.4-20220421
    argocd.argoproj.io/instance: apt-cacher-ng
    helm.sh/chart: apt-cacher-ng-0.0.5
    implementation: kube-vip
  name: apt-cacher-ng
  namespace: apt-cacher-ng
  resourceVersion: "307411776"
  uid: 07d8d6e5-a608-40e5-bfbd-5097c02e4d10
spec:
  allocateLoadBalancerNodePorts: true
  clusterIP: 10.43.171.146
  clusterIPs:
  - 10.43.171.146
  externalTrafficPolicy: Cluster
  internalTrafficPolicy: Cluster
  ipFamilies:
  - IPv4
  ipFamilyPolicy: SingleStack
  ports:
  - name: http
    nodePort: 32669
    port: 3142
    protocol: TCP
    targetPort: 3142
  selector:
    app.kubernetes.io/instance: apt-cacher-ng
    app.kubernetes.io/name: apt-cacher-ng
  sessionAffinity: None
  type: LoadBalancer
status:
  loadBalancer:
    ingress:
    - ip: 192.168.10.243
      ports:
      - port: 3142
        protocol: TCP
```

**Comment by reefland** [comment]
> Did you see any logging if you create a service with predefined `spec.loadblancerIP`

Well, when the annotation was not working and I put the `spec.loadblancerIP` back previously, it logged:
```
"Event occurred" object="apt-cacher-ng/apt-cacher-ng" fieldPath="" kind="Service" apiVersion="v1" type="Normal" reason="LoadbalancerIP" message="192.168.10.241 -> 192.168.10.243"
```

Which was handy to let me know it was assigned the IP address I requested.

**Comment by lubronzhan** [comment]
> Did you remove the spec.loadblancerIP from the service
> Yes I did.

Yeah the log 
```
Not clear what this line is trying to tell me:
reason="LoadbalancerIP" message="192.168.10.243 -> "
```
Means you removes it. And the upstream cloud-provider only checks this field.

It depends on its own implementer to show log around change in annotation.
I can try add more loggings
